Batteries/Data/Vector/Basic.lean

```diff
@@ -38,283 +39,267 @@ macro_rules
+/-- Returns a vector of size `1` with element `v`. -/
+@[inline] def singleton (v : α) : Vector α 1 := ⟨#[v], rfl⟩
 
-/-- Returns a vector of size `1` with a single element `v` -/
-def singleton (v : α) : Vector α 1 := mkVector 1 v
-
-/--
-The Inhabited instance for `Vector α n` with `[Inhabited α]` produces a vector of size `n`
-with all its elements equal to the `default` element of type `α`
--/
 instance [Inhabited α] : Inhabited (Vector α n) where
   default := mkVector n default
 
-/-- The list obtained from a vector. -/
-def toList (v : Vector α n) : List α := v.toArray.toList
+/-- Get an element of a vector using a `Fin` index. -/
+@[inline] def get (v : Vector α n) (i : Fin n) : α :=
+  v.toArray.get (i.cast v.size_toArray.symm)
 
-/-- nth element of a vector, indexed by a `Fin` type. -/
-def get (v : Vector α n) (i : Fin n) : α := v.toArray.get <| i.cast v.size_eq.symm
+/-- Get an element of a vector using a `USize` index and a proof that the index is within bounds. -/
+@[inline] def uget (v : Vector α n) (i : USize) (h : i.toNat < n) : α :=
+  v.toArray.uget i (v.size_toArray.symm ▸ h)
 
-/-- Vector lookup function that takes an index `i` of type `USize` -/
-def uget (v : Vector α n) (i : USize) (h : i.toNat < n) : α := v.toArray.uget i (v.size_eq.symm ▸ h)
-
-/-- `Vector α n` instance for the `GetElem` typeclass. -/
 instance : GetElem (Vector α n) Nat α fun _ i => i < n where
-  getElem := fun x i h => get x ⟨i, h⟩
+  getElem x i h := get x ⟨i, h⟩
 
 /--
-`getD v i v₀` gets the `iᵗʰ` element of v if valid.
-Otherwise it returns `v₀` by default
+Get an element of a vector using a `Nat` index. Returns the given default value if the index is out
+of bounds.
 -/
-def getD (v : Vector α n) (i : Nat) (v₀ : α) : α := Array.getD v.toArray i v₀
+@[inline] def getD (v : Vector α n) (i : Nat) (default : α) : α := v.toArray.getD i default
 
-/--
-`v.back! v` gets the last element of the vector.
-panics if `v` is empty.
--/
-abbrev back! [Inhabited α] (v : Vector α n) : α := v[n - 1]!
+/-- The last element of a vector. Panics if the vector is empty. -/
+@[inline] def back! [Inhabited α] (v : Vector α n) : α := v.toArray.back!
 
-/--
-`v.back?` gets the last element `x` of the array as `some x`
-if it exists. Else the vector is empty and it returns `none`
--/
-abbrev back? (v : Vector α n) : Option α := v[n - 1]?
+/-- The last element of a vector, or `none` if the array is empty. -/
+@[inline] def back? (v : Vector α n) : Option α := v.toArray.back?
 
-/-- Abbreviation for the last element of a non-empty `Vector`.-/
-abbrev back (v : Vector α (n + 1)) : α := v[n]
+/-- The last element of a non-empty vector. -/
+@[inline] def back (v : Vector α (n + 1)) : α :=
+  -- TODO: change to just `v[n]`
+  have : Inhabited α := ⟨v[n]⟩
+  v.back!
 
-/-- `Vector.head` produces the head of a vector -/
-abbrev head (v : Vector α (n+1)) := v[0]
+/-- The first element of a non-empty vector.  -/
+@[inline] def head (v : Vector α (n+1)) := v[0]
```

Comment:
Similarly here.

Reply:
There seems to be missing a `(n : Nat) [NeZero n] : 0 < n` simp lemma somewhere?